framework/src/modules/base_store.ts

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright © 2022 Lisk Foundation
+ *
+ * See the LICENSE file at the top-level directory of this distribution
+ * for licensing information.
+ *
+ * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
+ * no part of this software, including this file, may be copied, modified,
+ * propagated, or distributed except according to the terms contained in the
+ * LICENSE file.
+ *
+ * Removal or modification of this copyright notice is prohibited.
+ */
+import { Schema } from '@liskhq/lisk-codec';
+import { utils } from '@liskhq/lisk-cryptography';
+import { IterateOptions } from '@liskhq/lisk-db';
+import { ImmutableSubStore, SubStore } from '../state_machine/types';
+
+export interface ImmutableStoreGetter {
+	getStore: (moduleID: Buffer, storePrefix: Buffer) => ImmutableSubStore;
+}
+
+export interface StoreGetter {
+	getStore: (moduleID: Buffer, storePrefix: Buffer) => SubStore;
+}
+
+export abstract class BaseStore<T> {
+	private readonly _version: number;
+	private readonly _storePrefix: Buffer;
+	private readonly _subStorePrefix: Buffer;
+
+	public abstract schema: Schema;
+
+	public get storePrefix(): Buffer {
+		return this._storePrefix;
+	}
+
+	public get subStorePrefix(): Buffer {
+		return this._subStorePrefix;
+	}
+
+	public get key(): Buffer {
+		return Buffer.concat([this._storePrefix, this._subStorePrefix]);
+	}
+
+	public get name(): string {
+		const name = this.constructor.name.replace('Store', '');
+		return name.charAt(0).toLowerCase() + name.substr(1);
+	}
+
+	public constructor(moduleName: string, version = 0) {
+		this._version = version;
+		this._storePrefix = utils.hash(Buffer.from(moduleName, 'utf-8')).slice(0, 4);
+		// eslint-disable-next-line no-bitwise
+		this._storePrefix[0] &= 0x7f;
+		const versionBuffer = Buffer.alloc(2);
+		versionBuffer.writeUInt16BE(this._version, 0);
+		this._subStorePrefix = utils
+			.hash(Buffer.concat([Buffer.from(this.name, 'utf-8'), versionBuffer]))
+			.slice(0, 2);
```

Comment:
maybe we can extract all this logic to separate funcs so if we have to change anything we only have to change it in one file, same for other id, prefix, subStore prefix logics

Reply:
since it's already too big, ill create another issue